.github/actions/retry/src/action.ts

```diff
@@ -0,0 +1,129 @@
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS-IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+/**
+ * @fileoverview Github action implementation.
+ */
+import { error, info, debug } from "@actions/core";
+
+import { Inputs } from "./inputs";
+import { spawn } from "child_process";
+import ms from "milliseconds";
+import kill from "tree-kill";
+
+interface CommandResult {
+  success: boolean;
+  exitCode: number;
+  output: string[];
+}
+
+export class Action {
+  private OS = process.platform;
+  private getExecutable(): string {
+    return this.OS === "win32" ? "powershell" : "bash";
+  }
+
+  private async wait(ms: number) {
+    return new Promise((r) => setTimeout(r, ms));
+  }
+
+  private async runCommand(inputs: Inputs): Promise<CommandResult> {
+    const end_time = Date.now() + ms.hours(5);
+    const executable = this.getExecutable();
+
+    // Timeout exit code - 124
+    let exitCode = 124;
+    let done = false;
+    let output: string[] = [];
+
+    debug(`Running command ${inputs.command} on ${this.OS} using shell ${executable}`);
+    const child = spawn(inputs.command, { shell: executable });
+
+    child.stdout?.on("data", (data) => {
+      process.stdout.write(data);
+      output.push(data);
+    });
+    child.stderr?.on("data", (data) => {
+      process.stdout.write(data);
+      output.push(data);
+    });
+
+    child.on("exit", (code) => {
+      debug(`Code: ${code}`);
+
+      if (code === null) {
+        error("exit code cannot be null");
+        exitCode = -1;
+        return;
+      }
+
+      exitCode = code;
+      done = true;
+    });
+
+    do {
+      const polling_period = ms.seconds(1);
+      await this.wait(polling_period);
+    } while (Date.now() < end_time && !done);
+
+    debug('Command has finished.');
+
+    // timeout
+    if (!done && child.pid) {
+      kill(child.pid);
+    }
+
+    return {
+      success: exitCode === 0,
+      exitCode,
+      output,
+    };
+  }
+
+  private hasFlakyOutput(
+    flaky_test_output_lines: string[],
+    output: string[]
+  ): boolean {
+    const flaky_indicator = flaky_test_output_lines.find((flakyLine) =>
+      output.some((outputLine) => outputLine.includes(flakyLine))
+    );
+    if (flaky_indicator === undefined) {
+      info("Output doesn't contain flaky lines");
+      return false;
+    }
+    info(`Output contains flaky line: ${flaky_indicator}`);
+    return true;
+  }
+
+  public async runAction(inputs: Inputs): Promise<number> {
```

Comment:
Will earlier failures still show up in the logs and be noted? Ideally in the final log we say something like "this is attempt 2 of 2, find attempt 1 here" or have some indication that a rerun happened.

Reply:
Earlier failures will show up, yes. About retries - we can write whatever we want.

> find attempt 1 here

I'm not sure it's possible - to add a link to the start of the first attempt. It's a single log file, we can indicate that "this is attempt number 2".